### PR TITLE
Adds Eyecharts

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

Adds Eye charts

You can examine them to get a rough estimate of your eye health

image

I mapped them onto Meta, Delta, and Ice

## Why It's Good For The Game

It's peak for doctor RP. and more flavor

## Changelog

:cl: Melbert
add: Eyecharts to Medbay
/:cl:

### CODE REVIEW

**Comment by DaCoolBoss** [comment]
Add to medical lathe?

**Comment by Dawnseer** [comment]
Man what is this medical renaissance, so many fun new tools to use

**Comment by Loafin34** [comment]
now we need a ticket counter in medbay so you can go in when its' your ticket

**Comment by MrMelbert** [comment]
> Add to medical lathe?

You can make it from any plastic sign. Or you should be able to, I think

**Review by Wallemations — APPROVED** [review]

File: code/game/objects/structures/signs/sign_eyechart.dm [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		return
+
+	var/report = ""
+	if(!user.can_read(src, READING_CHECK_LITERACY, silent = TRUE) || !user.has_language(/datum/language/common, UNDERSTOOD_LANGUAGE))
+		if(user.is_blind())
+			return // you are hopeless
+		report = span_warning("You don't actually know what any of the symbols mean.")
+
+	else if(user.is_blind())
+		report = span_notice("Yes, feels like... \"E, P, D...\" Good thing this chart has braille!")
```
Comment (CRITAWAKETS) [inline code/game/objects/structures/signs/sign_eyechart.dm:21]: Doesn't actually check if you have any arms.

Comment (MrMelbert) [reply code/game/objects/structures/signs/sign_eyechart.dm:21]: You can't examine w/o arms so it's asserted

**Comment by DrAmazing343** [comment]
This is exactly the type of little detail that attracted me to the game in the first place. Bless you for never letting up on quality, Melbert.

**Comment by DaCoolBoss** [comment]
Should this have a range and/or LoS check so it can't be used through cameras? Unless we want this to be used through cameras?

**Comment by MrMelbert** [comment]
> Unless we want this to be used through cameras?

Sounds funny

File: code/game/objects/structures/signs/sign_eyechart.dm [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+		var/little_bad = isnull(eye) ? 20 : eye.low_threshold
+		var/very_bad = isnull(eye) ? 30 : eye.high_threshold
+
+		eye_goodness += ((get_dist(user, src) - 2) * 5) // add a modifier based on distance, so closer = "better", further = "worse"
+		if(user.has_status_effect(/datum/status_effect/eye_blur))
+			eye_goodness = max(eye_goodness, very_bad + 1)
+		if(user.is_nearsighted_currently())
+			eye_goodness = max(eye_goodness, little_bad + 1)
+
+		if(eye_goodness <= 0)
```
Comment (Ghommie) [inline code/game/objects/structures/signs/sign_eyechart.dm:40]: A switch() would be good here.

Comment (MrMelbert) [reply code/game/objects/structures/signs/sign_eyechart.dm:40]: Wouldn't work, non-constant values

File: code/game/objects/structures/signs/sign_eyechart.dm [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+			report = span_notice("\"E, F, P...\" Yep, you can read down to the green line.")
```
Comment (Isratosh) [inline code/game/objects/structures/signs/sign_eyechart.dm:41]: red line is 20/20 vision, green line is usually 20/30

does this check for worn glasses?

File: code/game/objects/structures/signs/sign_eyechart.dm (hunk 2 above)
Comment (Ghommie) [reply code/game/objects/structures/signs/sign_eyechart.dm:40]: Oh, ok.

File: code/game/objects/structures/signs/sign_eyechart.dm (hunk 3 above)
Comment (MrMelbert) [reply code/game/objects/structures/signs/sign_eyechart.dm:41]: Good shout

And yes, it checks `is_nearsighted_currently`

File: code/game/objects/structures/signs/sign_eyechart.dm [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+		if(eye_goodness <= 0)
+			report = span_notice("\"E, F, P...\" Yep, you can read down to the [colorblind ? "brown - wait, isn't it supposed to be red? -" : "red"] line.")
+		else if(eye_goodness < little_bad)
+			report = span_notice("\"E, F, P...\" You can make out most of the letters, but it gets a bit difficult past the [colorblind ? "grey - wait, isn't it supposed to be green? -" : "green"] line.")
+		else if(eye_goodness < very_bad)
+			report = span_warning("\"E, F, P..?\" You can make out the big letters, but the smaller ones are a bit of a blur.")
+		else
+			report = span_warning("\"E, P, D..?\" You can hardly make out the big letters, let alone the smaller ones.")
+
+	. += "<hr>You read through the chart, for old time's sake.<br>[report]"
```
Comment (SmArtKar) [inline code/game/objects/structures/signs/sign_eyechart.dm:56]: this entire thing could really use early returns, add ``<hr>You read through the chart, for old time's sake.`` as a separate element and report in ``return . + "whatever the report is"``

**Review by SmArtKar — CHANGES_REQUESTED** [review]

**Pull request merged**